Add USB support for RT600 & LPC55S69

mmahadevan108 opened the pull request:
Tested using `samples/subsys/usb/hid`,  `samples/subsys/usb/hid-mouse`, `samples/subsys/usb/mass`

mmahadevan108 committed:
drivers: usb: Add support for NXP LPC USB controller

1. Add support for NXP LPC USB controller
2. Do not check the return value from the
   kUSB_DeviceControlRun command as not all SDK drivers
   return a value
3. Use the kUSB_DeviceControlPreSetDeviceAddress
   command to set device address

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
modules: mcux: Add support for LPC USB

Add support for LPC USB controller

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
dts: mxrt685: Add USB support

Add USB and USB RAM information

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
west.yml: update NXP HAL for LPC USB device driver

Update NXP HAL repo to pull in SDK LPC USB driver.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
soc: LPC55S69: Add USB support

1. Update soc.c file to add USB clock setup
2. Add a linker script file to move USB transfer
   buffer and controller buffers to USB RAM
3. Update Kconfig's to add USB support
4. Add zephyr_udc0 nodelabel

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
dts: lpc55s69: Add USB support

Add USB and USB RAM information

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

klaasdannen commented:
Why limited to those two MCUs? What about the LPC54S16 which is also part of Zephyr and that supports USB?

MaureenHelm commented on drivers/usb/device/usb_dc_mcux_lpcip3511.c:
```diff
@@ -0,0 +1,775 @@
+/*
+ * Copyright (c) 2018-2019, NXP
```
Update to include 2021

MaureenHelm commented on drivers/usb/device/usb_dc_mcux_lpcip3511.c:
```diff
@@ -0,0 +1,775 @@
+/*
+ * Endpoint absolute index calculation:
+ *
+ * MCUX EHCI USB device controller supports a specific
```
Update to LPCIP3511

MaureenHelm commented on drivers/usb/device/usb_dc_mcux_lpcip3511.c:
```diff
@@ -0,0 +1,775 @@
+static usb_ep_ctrl_data_t s_ep_ctrl[NUM_OF_EP_MAX];
+static usb_device_struct_t dev_data;
+
+/* LPCIP3511 device driver interface */
+static const usb_device_controller_interface_struct_t ip3511 = {
+		USB_DeviceLpc3511IpInit, USB_DeviceLpc3511IpDeinit, USB_DeviceLpc3511IpSend,
+		USB_DeviceLpc3511IpRecv, USB_DeviceLpc3511IpCancel, USB_DeviceLpc3511IpControl
+};
```
Would this interface abstraction make it possible to merge this driver with the existing MCUX EHCI driver?

mmahadevan108 replied:
I have merged the two drivers. I have also renamed `usb_dc_mcux_ehci.c`  to `usb_dc_mcux.c` to reflect this change.

mmahadevan108 committed:
modules: hal_nxp: Updates for LPC USB

Add the defines to add support for LPC USB

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
dts: usb: Add MCUX USB driver

This USB driver is used for both NXP MXRT and LPC SoC
families that use the MCUX SDK.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
drivers: usb: Rename NXP EHCI driver

Rename to usb_dc_mcux as this driver is used for both
EHCI and LPC USB controllers

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 commented:
Any guidance on this build failure
```
HEAD is now at c74be389 trusted-firmware-m: platform: lpcxpresso55s69: Update SDK
ERROR: update failed for project lvgl

```

MaureenHelm commented on modules/hal_nxp/usb/usb_dc_mcux.h:
```diff
@@ -25,6 +25,43 @@
+
+#endif
+
+/* big/little endian */
+#define SWAP2BYTE_CONST(n) ((((n)&0x00FFU) << 8U) | (((n)&0xFF00U) >> 8U))
+
+#if defined(CONFIG_BIG_ENDIAN)
+#define USB_SHORT_FROM_LITTLE_ENDIAN(n) SWAP2BYTE_CONST(n)
+#else
+#define USB_SHORT_FROM_LITTLE_ENDIAN(n) (n)
+#endif
```
Use helper functions in `include/sys/byteorder.h`

mmahadevan108 replied:
Thank you, I have made the change.

MaureenHelm reviewed (commented):
This looks much better after merging with the EHCI driver and will be easier to maintain, thanks. I tested `samples/subsys/usb/hid-mouse` on `mimxrt1064_evk` and `mimxrt685_evk_cm33`.

> Any guidance on this build failure
>
> ```HEAD is now at c74be389 trusted-firmware-m: platform: lpcxpresso55s69: Update SDK ERROR: update failed for project lvgl```

No, but it looks like there are merge conflicts that need to be resolved anyway, so can you try rebasing and see if the failure recurs?

mmahadevan108 committed:
soc: RT685: Add USB support

1. Update soc.c file to add USB clock setup
2. Add a linker script file to move USB transfer
   buffer and controller buffers to USB RAM
3. Update Kconfig's to add USB support
4. Add zephyr_udc0 nodelabel

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
soc: rt: Update USB config and device tree node names

The USB config name has changed to USB_MCUX and the
device tree node name has changed from nxp_kinetis_usbd
to nxp_mcux_usbd

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 commented:
> Why limited to those two MCUs? What about the LPC55S16 which is also part of Zephyr and that supports USB? Or is that limited due to hardware availability?

This driver should work on LPC55S16 as well. I do not have a board to test, I will be happy to review if you can submit a PR.

mmahadevan108 commented:
@jfischer-no This PR is adding USB device support on some NXP platforms. Could you kindly review and let me know if you have any feedback.

MaureenHelm commented:
Rebase the hal_nxp PR to fix the buildkite failure. See zephyrproject-rtos/hal_nxp#109

hakehuang reviewed (commented):
LPC and RT685 testing PASS. but for some samples which use endpoint up to 4(LPC55S69)/6(RT685), does not support.

mbolivar-nordic reviewed (approved):
binding looks fine
